secondarylib/src/main/scala/com/socrata/datacoordinator/secondary/SecondaryWatcher.scala

```diff
@@ -175,29 +178,52 @@ class SecondaryWatcher[CT, CV](universe: => Managed[SecondaryWatcher.UniverseTyp
     }
   }
 
-  private def completeSecondaryMoveJobs(u: Universe[CT, CV] with SecondaryManifestProvider with PlaybackToSecondaryProvider
-                                                        with SecondaryMoveJobsProvider,
-                                    job: SecondaryRecord): Unit = {
+  private def completeSecondaryMoveJobs(u: Universe[CT, CV] with SecondaryManifestProvider
+                                                            with PlaybackToSecondaryProvider
+                                                            with SecondaryMoveJobsProvider,
+                                        job: SecondaryRecord): Unit = {
     val moveJobs = u.secondaryMoveJobs
 
     val storeId = job.storeId
     val datasetId = job.datasetId
-    val movesToStore = moveJobs.jobsToStore(storeId, datasetId)
 
     def forLog(jobs: Seq[SecondaryMoveJob]): String = jobs.map(_.id).toString()
 
     if (job.pendingDrop) {
-      if (movesToStore.nonEmpty) {
-        // TODO: this is actually a race condition....
-        log.warn(s"Upon pending drop of dataset {} from store {} there are unexepected jobs moving to the store: {}",
-          datasetId.toString, storeId.toString, forLog(movesToStore))
-      }
+      // For pending drops when "completing" move jobs we acquire a collocation lock to avoid race conditions
+      // Note: the only currently automated process to mark a dataset for pending drop is here in the secondary
+      // watcher after a dataset has completed replication to a destination store in a move job.
+      try {
+        log.info("Attempting to acquire collocation lock for pending drop of dataset.")
+        if (collocationLock.acquire(collocationLockTimeoutMillis)) {
+          log.info("Acquired collocation lock for pending drop of dataset.")
+
+          val movesToStore = moveJobs.jobsToStore(storeId, datasetId)
+
+          if (movesToStore.nonEmpty) {
+            log.error("Upon pending drop of dataset {} from store {} there are unexepected jobs moving to the store: {}",
+              datasetId.toString, storeId.toString, forLog(movesToStore))
+            throw new Exception("Unexpected jobs moving dataset to store when it is pending drop!")
```

Comment:
How is this handled eventually?  drop should trump move?

Reply:
This shouldn't happen, assuming I have reasoned about things correctly.... I guess if this does start happening the dataset will eventually end-up broken and we can figure out what is wrong and fix it.